xds/src/test/java/io/grpc/xds/XdsClientImplTest.java

```diff
@@ -0,0 +1,1369 @@
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new AssertionError(e);
+        }
+      });
+  private final FakeClock fakeClock = new FakeClock();
+
+  private final Queue<StreamObserver<DiscoveryResponse>> responseObservers = new ArrayDeque<>();
+  private final Queue<StreamObserver<DiscoveryRequest>> requestObservers = new ArrayDeque<>();
+
+  @Mock
+  private AggregatedDiscoveryServiceImplBase mockedDiscoveryService;
+  @Mock
+  private BackoffPolicy.Provider backoffPolicyProvider;
+  @Mock
+  private BackoffPolicy backoffPolicy1;
+  @Mock
+  private BackoffPolicy backoffPolicy2;
+  @Mock
+  private ConfigWatcher configWatcher;
+
+  private ManagedChannel channel;
+  private XdsClientImpl xdsClient;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    when(backoffPolicyProvider.get()).thenReturn(backoffPolicy1, backoffPolicy2);
+    when(backoffPolicy1.nextBackoffNanos()).thenReturn(10L, 100L);
+    when(backoffPolicy2.nextBackoffNanos()).thenReturn(20L, 200L);
+
+    String serverName = InProcessServerBuilder.generateName();
+    AggregatedDiscoveryServiceImplBase serviceImpl = new AggregatedDiscoveryServiceImplBase() {
+      @Override
+      public StreamObserver<DiscoveryRequest> streamAggregatedResources(
+          final StreamObserver<DiscoveryResponse> responseObserver) {
+        responseObservers.offer(responseObserver);
+        @SuppressWarnings("unchecked")
+        StreamObserver<DiscoveryRequest> requestObserver = mock(StreamObserver.class);
+        requestObservers.offer(requestObserver);
+        Answer<Void> closeRpc = new Answer<Void>() {
+          @Override
+          public Void answer(InvocationOnMock invocation) {
+            responseObserver.onCompleted();
+            return null;
+          }
+        };
+        doAnswer(closeRpc).when(requestObserver).onCompleted();
+
+        return requestObserver;
+      }
+    };
+    mockedDiscoveryService =
+        mock(AggregatedDiscoveryServiceImplBase.class, delegatesTo(serviceImpl));
+
+    cleanupRule.register(
+        InProcessServerBuilder
+            .forName(serverName)
+            .addService(mockedDiscoveryService)
+            .directExecutor()
+            .build()
+            .start());
+    channel =
+        cleanupRule.register(InProcessChannelBuilder.forName(serverName).directExecutor().build());
+    xdsClient =
+        new XdsClientImpl(channel, NODE, syncContext, fakeClock.getScheduledExecutorService(),
+            backoffPolicyProvider, fakeClock.getStopwatchSupplier().get());
+    // Only the connection to management server is established, no RPC request is sent until at
+    // least one watcher is registered.
+    assertThat(responseObservers).isEmpty();
+    assertThat(requestObservers).isEmpty();
+  }
+
+  @After
+  public void tearDown() {
+    xdsClient.shutdownRpcStream();
+    channel.shutdown();
```

Comment:
Actually I made the separation for control of `Channel` and `XdsClient` instances on purpose. Same idea as making a constructor for taking in a `Channel`. Now that we had the construction separates the control for `Channel` and `XdsClient`, the cleanup should do the same. `shutdownRpcStream()` is sort of `shutdown()` for testing, which does not shut down the channel. We shut down the channel separately, as we created it separately in setup. This won't confuse the reader by "creating a channel outside but shut it down inside".


Added check for pending tasks.

Reply:
Then you should have a test method to test `shutdown()` and verify channel is shutdown. Because your `shutdown()` implementation is shutting down the channel.

Reply:
We aren't aiming for 100% coverage right? I would say that's a fairly obvious thing and do not bother explicitly verify that.

Reply:
User will rely on the behavior `xdsClient.shutdown()`, and you don't test the behavior. User will not call `xdsClient.shutdownRpcStream()`

Reply:
We know what is `shutdown()` doing very clearly, by just taking a one-second glance of that method. And we know for sure it shuts down the channel in `shutdown()`. This is a trivial thing to be tested. Just like you have one constructor for testing only and the one used in execution calls it, but your test never need to cover this call chain.

Reply:
We should test behavior regardless of the implementation of the behavior is clear by a one-second glance. The test never cover the other constructor in your example is not about covering behavior.

Reply:
Done.